### PR TITLE
user pool

### PR DESCRIPTION
**Pull request by RuslanNikolaev**
added: user pool, behavior of registration buttons, command /showpool

### CODE REVIEW

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java [hunk 1]
```diff
@@ -5,26 +5,45 @@
...
         return Button.REGISTER;
     }
 
     @Override
     public CommandResult execute(Update update) {
-        return CommandResult.builder()
-                .replyMessage("You are registered for upcoming game!")
-                .keyboardMarkup(keyboardFactory.keyboardOf(UNREGISTER))
-                .build();
+        synchronized (userPool) {
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java]: I think we can achieve more fine grained sync. Will discuss this

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java [hunk 2]
```diff
@@ -47,6 +47,6 @@ private String userNameFromUpdate(Update update) {
     }
 
     private String toUserName(User user) {
-        return user.getFirstName() + " " + user.getLastName();
+        return user.getFirstName() + (user.getLastName() != null ? " " + user.getLastName() : "");
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java]: [future_notes] This should be moved to some kind of UserOperations service/util class, encapsulate user-specific logic there. Let's leave it as is for now.

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    @Override
+    public CommandResult execute(Update update) {
+        if(userPool.isEmptyPool()) {
+            return CommandResult.builder()
+                    .replyMessage("No registered users. You can be first!")
+                    .keyboardMarkup(keyboardFactory.keyboardOf(REGISTER))
+                    .build();
+        }
+        return CommandResult.builder()
+                .replyMessage(userPool.toString())
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java]: [future_notes] UserPool should return _immutable_ pool data that could be converted to command reply message somewhere else.

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.stereotype.Component;
+import org.telegram.telegrambots.api.objects.User;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Component
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Define bean in java config.

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+public class UserPool {
+
+    private static final int POOL_SIZE = 4;
+    public static long POOL_REQUEST_TIMEOUT = TimeUnit.MINUTES.toMillis(15);
+    private long lastPoolUpdate;
+
+    @Getter private Map<Integer, User> userPool = new HashMap<>();
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: - Let's think of better encapsulation/synchronization.
- Take existing expirable map implementation so we wouldn't rely on last timestamp and so on

**Review by jnopnop — CHANGES_REQUESTED** [review]

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java (hunk 1 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java]: Okay) Removed for a while

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java (hunk 2 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java]: Moved to the UserUtils class

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java (hunk 3 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java]: Function is renamed and implemented in the UserUtils class

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java (hunk 4 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Done

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java (hunk 5 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Now used the ExpiringMap with https://github.com/jhalterman/expiringmap. 
He allows remove the timestamp implementation

File: pom.xml [hunk 6]
```diff
@@ -66,6 +66,11 @@
 			<artifactId>lombok</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>net.jodah</groupId>
+			<artifactId>expiringmap</artifactId>
+		  	<version>0.5.8</version>
```
Comment (jnopnop) [inline pom.xml]: pull this to properties section

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java [hunk 7]
```diff
@@ -4,22 +4,24 @@
...
 
 @AllArgsConstructor
 public class HelloCommand implements TextCommand {
 
     private static final Pattern CMD_PATTERN = Pattern.compile("/hello");
-    private static final String REPLY = "Hello, %s!";
+    private static final String REPLY = "Hello, %s";
     private static final String UNKNOWN_USER = "anonymous";
 
+    private static final UserUtils USER_UTILS = UserUtils.getINSTANCE();
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java]: Say "no" to singleton. Use either spring bean with corresponding scope, which is default, or simply static class. Later is preferred in given case.

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    private String getMessage() {
+        return userPool.isEmpty() ? "No registered users. You can be first!"
+                : USER_UTILS.getPoolList(userPool.getUserPool());
+    }
+
+    private InlineKeyboardMarkup getButton() {
+        return userPool.isEmpty() ? keyboardFactory.keyboardOf(REGISTER)
+                : keyboardFactory.keyboardOf(START);
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java:43]: [future] Let's leave it as is for now, but I'm thinking of auto-starting the game after pool is filled out.

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.concurrent.TimeUnit;
+
+public class UserPool {
+
+    private static final int POOL_SIZE = 4;
+
+    @Getter private Map<Integer, User> userPool = ExpiringMap.builder()
+            .expiration(15, TimeUnit.MINUTES)
+            .expirationPolicy(ExpirationPolicy.ACCESSED)
+            .maxSize(4)
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Use constant defined above

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class UserPool {
+
+    private static final int POOL_SIZE = 4;
+
+    @Getter private Map<Integer, User> userPool = ExpiringMap.builder()
+            .expiration(15, TimeUnit.MINUTES)
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Move expiration to properties

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import org.telegram.telegrambots.api.objects.User;
+
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class UserPool {
+
+    private static final int POOL_SIZE = 4;
+
+    @Getter private Map<Integer, User> userPool = ExpiringMap.builder()
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Do not expose this map to outer world as someone could change its contents. Make explicit getter which returns copy of the map e.g. ImmutableMap.copyOf(userPool)

File: src/main/java/org/dormitory/autobotsoccub/user/UserUtils.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+import lombok.Getter;
+import org.telegram.telegrambots.api.objects.User;
+
+import javax.inject.Singleton;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Singleton
+public class UserUtils {
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/user/UserUtils.java]: It seems there is no adequate reason to make this class a bean. Let it be plain old utility class with static methods.
BTW, we're not using component scan, spring beans should be defined in corresponding contexts.

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java [hunk 13]
```diff
@@ -5,26 +5,42 @@
...
     public Button getCommandButton() {
         return Button.REGISTER;
     }
 
     @Override
     public CommandResult execute(Update update) {
+        User currentUser = update.getCallbackQuery().getFrom();
+        if(userPool.containsUser(currentUser.getId())) {
+            return CommandResult.builder()
+                    .replyMessage(currentUser.getFirstName() + " you are already registered!")
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java:31]: [future] this message is gonna spam the whole chat, empty reply might be better.

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java [hunk 14]
```diff
@@ -5,26 +5,42 @@
...
+        if (userPool.isFull()) {
+            return CommandResult.builder()
+                    .replyMessage(currentUser.getFirstName() + " are registered for upcoming game!\nThe game begins...")
+                    .keyboardMarkup(keyboardFactory.keyboardOf(START))
+                    .build();
+        }
         return CommandResult.builder()
-                .replyMessage("You are registered for upcoming game!")
-                .keyboardMarkup(keyboardFactory.keyboardOf(UNREGISTER))
+                .replyMessage(currentUser.getFirstName() + " are registered for upcoming game!")
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java:43]: [future] the same here

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/UnregisterCommand.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        return Button.UNREGISTER;
+    }
+
+    @Override
+    public CommandResult execute(Update update) {
+        User currentUser = update.getCallbackQuery().getFrom();
+        if (userPool.containsUser(currentUser.getId())) {
+            userPool.removeUser(currentUser);
+            return CommandResult.builder()
+                    .replyMessage(currentUser.getFirstName() + ", your registration for the game is canceled! But you can change your mind...")
```
Comment (jnopnop) [inline src/main/java/org/dormitory/autobotsoccub/command/querycommand/UnregisterCommand.java:32]: [future] the same as for register command; we gonna decrease unnecessary level of verbosity

**Review by jnopnop — CHANGES_REQUESTED** [review]
Needs some cosmetic changes. The logic is good.

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java (hunk 8 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/txtcommand/ShowPoolCommand.java:43]: ofc, in the future we will do auto-starting

File: pom.xml (hunk 6 above)
Comment (RuslanNikolaev) [reply pom.xml]: done

File: src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java (hunk 7 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/txtcommand/HelloCommand.java]: I say: "No, singleton!", and he left) Now used static.

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java (hunk 11 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Okay, its done

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java (hunk 10 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Moved, default value = 15

File: src/main/java/org/dormitory/autobotsoccub/user/UserPool.java (hunk 9 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserPool.java]: Exactly! Thanks

**Review by jnopnop — APPROVED** [review]

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java (hunk 13 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java:31]: With an empty response will be thrown telegtamApiExeption(also with empty message or space), so I propose to leave it as it is

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java (hunk 14 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/querycommand/RegisterCommand.java:43]: ok, in the future we will investigate this moment

File: src/main/java/org/dormitory/autobotsoccub/command/querycommand/UnregisterCommand.java (hunk 15 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/command/querycommand/UnregisterCommand.java:32]: The same here)

File: src/main/java/org/dormitory/autobotsoccub/user/UserUtils.java (hunk 12 above)
Comment (RuslanNikolaev) [reply src/main/java/org/dormitory/autobotsoccub/user/UserUtils.java]: Good, now it is old utility class with static methods.

**Pull request merged**